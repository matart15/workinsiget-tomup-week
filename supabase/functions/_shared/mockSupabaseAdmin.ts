const SUPABASE_FUNCTION_NAMES = [
  "from",
  "select",
  "eq",
  "delete",
  "update",
  "insert",
  "limit",
  "single",
] as const;
type ReturnFunctionName = (typeof SUPABASE_FUNCTION_NAMES)[number];

export const mockSupabaseAdmin: any = (
  returnValues: Record<ReturnFunctionName, any>[],
) => {
  let allCalls: any = [];
  let allArgs: any = [];
  let currentChainIndex = 0;

  const supabaseAdmin: any = {};
  for (const functionName of SUPABASE_FUNCTION_NAMES) {
    supabaseAdmin[functionName] = (...args: any[]) => {
      const returnValue = returnValues[currentChainIndex];
      const returnFunctionName: ReturnFunctionName = Object.keys(
        returnValue,
      )[0] as any;

      allArgs.push({ [functionName]: args });
      if (returnFunctionName === functionName) {
        allCalls.push(allArgs);
        currentChainIndex++;
        return returnValue[returnFunctionName];
      }
      return supabaseAdmin;
    };
  }

  return {
    allCalls,
    supabaseAdmin,
  };
};
