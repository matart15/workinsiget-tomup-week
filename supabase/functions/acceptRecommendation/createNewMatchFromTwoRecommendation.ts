import { SupabaseClient } from "https://esm.sh/@supabase/supabase-js@2";

import { Database } from "../_shared/database.ts";

export async function createNewMatchFromTwoRecommendation({
  recommendation_id,
  supabaseAdmin,
  otherUserId,
  currentUserId,
  otherUserRecommendationId,
}: {
  recommendation_id: number;
  supabaseAdmin: SupabaseClient<Database>;
  otherUserId: string;
  currentUserId: string;
  otherUserRecommendationId: number;
}) {
  const { error: matchError } = await supabaseAdmin.from("u_matches").insert({
    user1_id: currentUserId,
    user2_id: otherUserId,
  });

  if (matchError) {
    throw matchError;
  }

  //remove recommendation
  const { error: deleteRecommendationError } = await supabaseAdmin
    .from("u_recommendations")
    .delete()
    .eq("id", recommendation_id);

  if (deleteRecommendationError) {
    throw deleteRecommendationError;
  }
  // remove other user recommendation
  const { error: deleteOtherUserRecommendation } = await supabaseAdmin
    .from("u_recommendations")
    .delete()
    .eq("id", otherUserRecommendationId);

  if (deleteOtherUserRecommendation) {
    throw deleteOtherUserRecommendation;
  }
}
