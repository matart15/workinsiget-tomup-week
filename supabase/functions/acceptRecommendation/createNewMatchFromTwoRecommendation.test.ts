import { assertRejects } from "https://deno.land/std@0.206.0/testing/asserts.ts";
import { assertSnapshot } from "https://deno.land/std@0.206.0/testing/snapshot.ts";

import { mockSupabaseAdmin } from "../_shared/mockSupabaseAdmin.ts";
import { createNewMatchFromTwoRecommendation } from "./createNewMatchFromTwoRecommendation.ts";

Deno.test(
  "createNewMatchFromTwoRecommendation should create a new match and delete the recommendations",
  async (t) => {
    const recommendation_id = 123;
    const otherUserRecommendationId = 456;
    const currentUserId = "789";
    const otherUserId = "012";
    const { supabaseAdmin, allCalls } = mockSupabaseAdmin([
      {
        insert: {
          data: {},
        },
      },
      {
        eq: {
          data: {},
        },
      },
      {
        eq: {
          data: {},
        },
      },
    ]);

    // Act
    await createNewMatchFromTwoRecommendation({
      recommendation_id,
      supabaseAdmin,
      otherUserId,
      currentUserId,
      otherUserRecommendationId,
    });

    // Assert
    await assertSnapshot(t, allCalls);
  },
);

Deno.test(
  "createNewMatchFromTwoRecommendation should throw an error if match insertion fails",
  async () => {
    const recommendation_id = 123;
    const otherUserRecommendationId = 456;
    const currentUserId = "789";
    const otherUserId = "012";
    const { supabaseAdmin } = mockSupabaseAdmin([
      {
        insert: {
          error: new Error("Match insertion failed"),
        },
      },
    ]);

    // Act and Assert
    await assertRejects(
      async () => {
        await createNewMatchFromTwoRecommendation({
          recommendation_id,
          supabaseAdmin,
          otherUserId,
          currentUserId,
          otherUserRecommendationId,
        });
      },
      Error,
      "Match insertion failed",
    );
  },
);

Deno.test(
  "createNewMatchFromTwoRecommendation should throw an error if recommendation deletion fails",
  async () => {
    const recommendation_id = 123;
    const otherUserRecommendationId = 456;
    const currentUserId = "789";
    const otherUserId = "012";
    const { supabaseAdmin } = mockSupabaseAdmin([
      {
        insert: {
          data: {},
        },
      },
      {
        eq: {
          error: new Error("Recommendation deletion failed"),
        },
      },
    ]);

    // Act and Assert
    await assertRejects(
      async () => {
        await createNewMatchFromTwoRecommendation({
          recommendation_id,
          supabaseAdmin,
          otherUserId,
          currentUserId,
          otherUserRecommendationId,
        });
      },
      Error,
      "Recommendation deletion failed",
    );
  },
);

Deno.test(
  "createNewMatchFromTwoRecommendation should throw an error if other user recommendation deletion fails",
  async () => {
    const recommendation_id = 123;
    const otherUserRecommendationId = 456;
    const currentUserId = "789";
    const otherUserId = "012";
    const { supabaseAdmin } = mockSupabaseAdmin([
      {
        insert: {
          data: {},
        },
      },
      {
        eq: {
          data: {},
        },
      },
      {
        eq: {
          error: new Error("Other user recommendation deletion failed"),
        },
      },
    ]);

    // Act and Assert
    await assertRejects(
      async () => {
        await createNewMatchFromTwoRecommendation({
          recommendation_id,
          supabaseAdmin,
          otherUserId,
          currentUserId,
          otherUserRecommendationId,
        });
      },
      Error,
      "Other user recommendation deletion failed",
    );
  },
);
