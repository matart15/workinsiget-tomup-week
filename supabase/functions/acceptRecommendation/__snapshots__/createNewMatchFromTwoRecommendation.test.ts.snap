export const snapshot = {};

snapshot[`createNewMatchFromTwoRecommendation should create a new match and delete the recommendations 1`] = `
[
  [
    {
      from: [
        "u_matches",
      ],
    },
    {
      insert: [
        {
          user1_id: "789",
          user2_id: "012",
        },
      ],
    },
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      delete: [],
    },
    {
      eq: [
        "id",
        123,
      ],
    },
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      delete: [],
    },
    {
      eq: [
        "id",
        456,
      ],
    },
  ],
  [
    {
      from: [
        "u_matches",
      ],
    },
    {
      insert: [
        {
          user1_id: "789",
          user2_id: "012",
        },
      ],
    },
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      delete: [],
    },
    {
      eq: [
        "id",
        123,
      ],
    },
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      delete: [],
    },
    {
      eq: [
        "id",
        456,
      ],
    },
  ],
  [
    {
      from: [
        "u_matches",
      ],
    },
    {
      insert: [
        {
          user1_id: "789",
          user2_id: "012",
        },
      ],
    },
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      delete: [],
    },
    {
      eq: [
        "id",
        123,
      ],
    },
    {
      from: [
        "u_recommendations",
      ],
    },
    {
      delete: [],
    },
    {
      eq: [
        "id",
        456,
      ],
    },
  ],
]
`;
